lib/std/gpu.zig

```diff
@@ -67,6 +79,58 @@ pub fn fragmentDepth(comptime ptr: *addrspace(.output) f32) void {
     );
 }
 
+/// Establish an struct type as a memory interface block.
+pub fn block(comptime T: type) void {
+    asm volatile (
+        \\OpDecorate %T Block
+        :
+        : [T] "" (T),
```

Comment:
So again for future reference, I don't feel good about the way that this is done because it applies a side effect to the type at compile time. For example, stuff like this will produce an invalid module, as it will apply the decoration twice:
```zig
const A = ...
std.gpu.block(A);
std.gpu.block(A);
```
This is a bigger problem with stuff like
```
const A = [3]u32;
std.gpu.arrayStride(A, 10);

// In another source file
const B = [3]u32;
std.gpu.arrayStride(B, 20);
```
In the compiler internals, there is only one `[3]u32` type, and so that type only gets emitted once, but the decoration gets applied multiple times and so will produce something like
```llvm
           %u32 = OpTypeInt 32 0
             %3 = OpConstant %u32 3
%array_of_3_u32 = OpTypeArray %u32 %3
%OpDecorate %array_of_3_u32 ArrayStride 10
%OpDecorate %array_of_3_u32 ArrayStride 20
```
I'm not sure what the right way to solve this is: There two most straightforward ways are either extending the Zig language, or using inline assembly. The most aesthetically pleasing is something like `[3]stride(10) u32` of course, but I'm not sure about extending the language for every little feature that SPIR-V may have to offer. There are [quite a lot](https://registry.khronos.org/SPIR-V/specs/unified1/SPIRV.html#Decoration), and they may evolve over time as well. As for inline assembly, one way would be to create types using inline assembly, but that would semantically probably return an opaque type, meaning that you couldn't do stuff like array indexing with a type created like this. Also it will be hard to create for example a matrix type and use that in a structure, because Zig considers it opaque...

A number of these attributes can be derived by the compiler already. Even if structures are mostly logical in SPIR-V, they still have a particular memory layout if stored in global memory. This should be encoded in the compiler's ABI calculations which then translate into `OpMemberDecoration %T i Offset <calculated>`. We can do this today.

Reply:
After brief discussion with @alichraghi we reached the conclusion that ArrayStride can be represented by a "native" stride of `@sizeOf(Element)` and that non-native strides can be represented by wrapping the element in a struct with some extra padding. A similar method should cover the `Offset` decoration.